cpp/src/datetime/datetime_ops.cu

```diff
@@ -68,98 +67,152 @@ struct extract_component_operator {
+struct launch_functor {
   column_view input;
   mutable_column_view output;
 
-  launch_extract_component(column_view inp, mutable_column_view out)
+  launch_functor(column_view inp, mutable_column_view out)
       : input(inp), output(out) {}
 
   template <typename Element>
   typename std::enable_if_t<!cudf::is_timestamp_t<Element>::value, void>
-  operator()(cudaStream_t stream) {
+  operator()(cudaStream_t stream) const {
     CUDF_FAIL("Cannot extract datetime component from non-timestamp column.");
   }
 
   template <typename Timestamp>
   typename std::enable_if_t<cudf::is_timestamp_t<Timestamp>::value, void>
-  operator()(cudaStream_t stream) {
+  operator()(cudaStream_t stream) const {
     thrust::transform(rmm::exec_policy(stream)->on(stream),
                       input.begin<Timestamp>(), input.end<Timestamp>(),
-                      output.begin<int16_t>(),
-                      extract_component_operator<Timestamp, Component>{});
+                      output.begin<OutputColT>(),
+                      TransformFunctor{});
   }
 };
 
-template <datetime_component Component>
-std::unique_ptr<column> extract_component(column_view const& column,
+// Create an output column by applying the functor to every element from the input column
+template <typename TransformFunctor, cudf::type_id OutputColCudfT>
+std::unique_ptr<column> apply_datetime_op(column_view const& column,
                                           cudaStream_t stream,
                                           rmm::mr::device_memory_resource* mr) {
   auto size = column.size();
-  auto type = data_type{type_id::INT16};
+  auto output_col_type = data_type{OutputColCudfT};
   auto null_mask = copy_bitmask(column, stream, mr);
   auto output = std::make_unique<cudf::column>(
-      type, size, rmm::device_buffer{size * cudf::size_of(type), stream, mr},
+      output_col_type, size, rmm::device_buffer{size * cudf::size_of(output_col_type), stream, mr},
       null_mask, column.null_count(),
       std::vector<std::unique_ptr<cudf::column>>{});
 
-  auto launch = launch_extract_component<Component>{
-      column, static_cast<mutable_column_view>(*output)};
+  auto launch = launch_functor<TransformFunctor,
+                               typename cudf::experimental::id_to_type_impl<OutputColCudfT>::type>
+      {column, static_cast<mutable_column_view>(*output)};
 
   experimental::type_dispatcher(column.type(), launch, stream);
 
   return output;
 }
+
 }  // namespace detail
 
 std::unique_ptr<column> extract_year(column_view const& column,
                                      rmm::mr::device_memory_resource* mr) {
   CUDF_FUNC_RANGE();
-  return detail::extract_component<detail::datetime_component::YEAR>(column, 0,
-                                                                     mr);
+  return detail::apply_datetime_op<
+      detail::extract_component_operator<detail::datetime_component::YEAR>, cudf::INT16>
```

Comment:
This looks like it should use an actual type parameter instead of a `type_id` non-type parameter. 
```suggestion
      detail::extract_component_operator<detail::datetime_component::YEAR>, int16_t>
```

Reply:
the `type_id` is needed to create the  output column `data_type` and the underlying data type is required in the transform method (for the output iterator). i just chose the option to pass the `type_id` and use it to compute the underlying data type internally. if we pass the real data type, we have to still use that to compute the `type_id`. from what i could tell, converting one from another is required.